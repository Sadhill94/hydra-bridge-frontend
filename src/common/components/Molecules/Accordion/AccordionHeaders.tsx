import { useTranslation } from "react-i18next";

import {
  StyledAccordionHeader,
  AccordionIcon,
  StyledReceiveDetailsAccordionHeader,
  StyledAccordionReceiveDetailsGasFeeEstimation,
  StyledAccordionReceiveDetailsAmountOut,
} from "./styles";
import { FlexWrapper } from "../../Atoms/Wrappers/Wrapper";
import React, { ReactNode } from "react";
import Icon from "../../Icon/Icon";

export type AccordionHeaderProps = {
  children: ReactNode;
  isOpen?: boolean;
};
export const AccordionHeader = ({ isOpen, children }: AccordionHeaderProps) => {
  return (
    <StyledAccordionHeader>
      <FlexWrapper flexDirection={"row"}>
        {children}
        {/* TODO 44 integrate font awesome in next PR and replace */}
        <AccordionIcon
          name={"cutArrowDown"}
          width={"2.4rem"}
          height={"2.4rem"}
          style={{ marginLeft: "1.7rem" }}
          isopen={isOpen ? 1 : 0}
        />
      </FlexWrapper>
    </StyledAccordionHeader>
  );
};

type ReceiveDetailsAccordionHeaderProps = {
  isOpen?: boolean;
  amountOut?: string;
  inProgress?: boolean;
  isDisabled?: boolean;
  gasFees: string;
};
export const ReceiveDetailsAccordionHeader = ({
  isOpen,
  amountOut,
  inProgress,
  gasFees,
}: ReceiveDetailsAccordionHeaderProps) => {
  const { t } = useTranslation();

  return (
    <StyledReceiveDetailsAccordionHeader isOpen={isOpen}>
      <FlexWrapper flexDirection={"row"} style={{ padding: "1.1rem 1.6rem" }}>
        <FlexWrapper
          flexDirection={"row"}
          alignItems={"center"}
          justifyContent={"space-between"}
        >
          <StyledAccordionReceiveDetailsAmountOut>
            {amountOut ? (
              amountOut
            ) : inProgress ? (
              <span>{t("fetching-prices")}...</span>
            ) : (
              "0.0"
            )}
          </StyledAccordionReceiveDetailsAmountOut>
          <FlexWrapper
            flexDirection={"row"}
            justifyContent={"flex-end"}
            inlineFlex
          >
            {/* TODO loading spinner integration in next PR */}
            {!amountOut ? null : inProgress ? (
              ""
            ) : (
              <>
                {/* TODO 44 integrate font awesome in next PR and replace */}
                <Icon name={"copy"} size={"1.6rem"} />
                <StyledAccordionReceiveDetailsGasFeeEstimation>
                  {gasFees}
                </StyledAccordionReceiveDetailsGasFeeEstimation>
              </>
            )}
          </FlexWrapper>
        </FlexWrapper>

        {/* TODO 44 integrate font awesome in next PR and replace */}
        <AccordionIcon
          name={"cutArrowDown"}
          width={"2.4rem"}
          height={"2.4rem"}
          style={{ marginLeft: "1.7rem" }}
          isopen={isOpen ? 1 : 0}
        />
      </FlexWrapper>
    </StyledReceiveDetailsAccordionHeader>
  );
};
