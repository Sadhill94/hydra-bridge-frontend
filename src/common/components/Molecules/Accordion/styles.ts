import styled from "styled-components";

import Icon from "../../Icon/Icon";

import { stakenetTheme as theme } from "../../../../shell/theme/stakenetTheme";

export const AccordionContainer = styled.div`
  display: flex;
  tab-index: 0;
  flex-direction: column;
  justify-content: center;
  background-color: lightgrey;
  border-radius: 10px;
  height: auto;
  padding: 2%;
  width: 100%;
  text-align: center;
`;
export const AccordionContent = styled.div`
  width: 100%;
  padding-top: ${theme.margin.lg};
`;

export const AccordionIcon = styled(Icon)<{ isopen: boolean }>`
  transform: ${(props) => (props.isopen ? "rotate(90deg)" : "rotate(0)")};
  transition: transform 0.3s linear;
`;
