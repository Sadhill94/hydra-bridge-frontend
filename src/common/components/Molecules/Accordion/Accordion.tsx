import React, { ReactNode, useState } from "react";
import AnimateHeight from "react-animate-height";

import { AccordionContent, AccordionIcon } from "./styles";
import { ContainerCard } from "../../Atoms/Containers/Container";
import { FakeButton } from "../../Atoms/Buttons/Button";
import { FlexWrapper } from "../../Atoms/Wrappers/Wrapper";

export type AccordionProps = {
  header: ReactNode;
  content: ReactNode;
};

const Accordion = ({ header, content }: AccordionProps) => {
  const [height, setHeight] = useState<string | number>(0);

  const handleToggle = () => {
    setHeight(height === 0 ? "auto" : 0);
  };
  return (
    <ContainerCard padding={"0"}>
      <FakeButton
        ariaLabel={"open accordion"}
        aria-expanded={height !== 0}
        style={{ padding: "1.8rem 2.4rem" }}
        onClick={handleToggle}
      >
        <FlexWrapper flexDirection={"row"} style={{ pointerEvents: "none" }}>
          {header}
          <AccordionIcon
            name={"cutArrowRight"}
            width={"1.7rem"}
            height={"1.7rem"}
            style={{ marginLeft: "1.7rem" }}
            isopen={height !== 0 ? 1 : 0}
          />
        </FlexWrapper>
      </FakeButton>
      <AnimateHeight height={height} duration={500} style={{ width: "100%" }}>
        <AccordionContent>{content}</AccordionContent>
      </AnimateHeight>
    </ContainerCard>
  );
};

export default Accordion;
