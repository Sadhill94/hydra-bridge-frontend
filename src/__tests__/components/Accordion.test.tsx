import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Accordion, {
  AccordionProps,
} from "../../common/components/Molecules/Accordion/Accordion";
import { AccordionHeader } from "../../common/components/Molecules/Accordion/AccordionHeaders";
import { InputLabel as Label } from "../../common/components/Atoms/Label/Label";

describe("The Accordion", () => {
  const getHeader = (isOpen = false) => (
    <AccordionHeader isOpen={isOpen}>
      <Label>Foo</Label>
    </AccordionHeader>
  );
  const content = <>Bar</>;

  it("should render correctly when closed", async () => {
    const props: AccordionProps = { header: getHeader(), content };
    const { asFragment } = render(<Accordion {...props} />);
    const element = screen.getByRole("button");

    const icon = element.querySelector<SVGElement>(
      `[data-icon="cutArrowDown"]`
    );

    expect(asFragment()).toMatchSnapshot();
    expect(element.getAttribute("aria-expanded")).toBe("false");
    expect(icon!.hasAttribute("data-icon")).toBe(true);
    expect(icon!.getAttribute("data-icon")).toBe("cutArrowDown");
    expect(icon).toHaveStyle("transform: rotate(-90deg)");
  });

  it("should render correctly when open", async () => {
    const props: AccordionProps = { header: getHeader(true), content };

    const { asFragment } = render(<Accordion {...props} />);
    const element = screen.getByRole("button");
    fireEvent.click(element);
    const icon = element.querySelector<SVGElement>(
      `[data-icon="cutArrowDown"]`
    );

    expect(asFragment()).toMatchSnapshot();
    expect(element.getAttribute("aria-expanded")).toBe("true");
    expect(icon!.hasAttribute("data-icon")).toBe(true);
    expect(icon!.getAttribute("data-icon")).toBe("cutArrowDown");
    expect(icon).toHaveStyle("transform: rotate(0)");
  });

  describe("The Accordion independently", () => {
    const props: AccordionProps = { header: getHeader(), content };
    const onToggle = jest.fn();

    afterEach(() => {
      jest.clearAllMocks();
    });

    it("should open and emit the new status when onToggle is defined but not isOpenFromParent", async () => {
      const { asFragment } = render(
        <Accordion {...props} onToggle={onToggle} />
      );

      const element = screen.getByRole("button");
      fireEvent.click(element);

      expect(asFragment()).toMatchSnapshot();
      expect(onToggle).toHaveBeenCalledTimes(1);
      expect(onToggle).toHaveBeenCalledWith(true);
      expect(element.getAttribute("aria-expanded")).toBe("true");
    });

    it("should not open itself when is open is handled by parent", async () => {
      const { asFragment } = render(
        <Accordion {...props} onToggle={onToggle} isOpenFromParent={false} />
      );

      const element = screen.getByRole("button");
      fireEvent.click(element);

      expect(asFragment()).toMatchSnapshot();
      expect(onToggle).toHaveBeenCalledTimes(1);
      expect(onToggle).toHaveBeenCalledWith(true);
      expect(element.getAttribute("aria-expanded")).toBe("false");
    });
  });
});
