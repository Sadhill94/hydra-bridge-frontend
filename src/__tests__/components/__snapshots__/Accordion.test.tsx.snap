// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Accordion The Accordion independently should not open itself when is open is handled by parent 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL sc-jRQBWg fGXmcv hJXOBd"
    style="width: 100%;"
  >
    <div
      aria-expanded="false"
      aria-label="open accordion"
      class="sc-furwcr gPiovd"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-pVTFL yzYaz sc-jrQzAO gCBRHg"
      >
        <div
          class="sc-pVTFL ilAcon"
        >
          <label
            class="sc-ksdxgE fxGKky"
          >
            Foo
          </label>
          <svg
            class="sc-iqseJM eSaCCB"
            color="#FFFFFF"
            data-icon="cutArrowDown"
            height="2.4rem"
            isopen="0"
            style="margin-left: 1.7rem;"
            width="2.4rem"
          >
            cut-arrow-down.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="rah-static rah-static--height-zero"
      style="width: 100%; height: 0px; overflow: hidden;"
    >
      <div
        style="display: none;"
      >
        Bar
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`The Accordion The Accordion independently should open and emit the new status when onToggle is defined but not isOpenFromParent 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL sc-jRQBWg fGXmcv hJXOBd"
    style="width: 100%;"
  >
    <div
      aria-expanded="true"
      aria-label="open accordion"
      class="sc-furwcr gPiovd"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-pVTFL yzYaz sc-jrQzAO gCBRHg"
      >
        <div
          class="sc-pVTFL ilAcon"
        >
          <label
            class="sc-ksdxgE fxGKky"
          >
            Foo
          </label>
          <svg
            class="sc-iqseJM eSaCCB"
            color="#FFFFFF"
            data-icon="cutArrowDown"
            height="2.4rem"
            isopen="0"
            style="margin-left: 1.7rem;"
            width="2.4rem"
          >
            cut-arrow-down.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="rah-animating rah-animating--down rah-animating--to-height-auto"
      style="width: 100%; height: 0px; overflow: hidden; transition: height 500ms ease 0ms;"
    >
      <div
        style=""
      >
        Bar
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`The Accordion should render correctly when closed 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL sc-jRQBWg fGXmcv hJXOBd"
    style="width: 100%;"
  >
    <div
      aria-expanded="false"
      aria-label="open accordion"
      class="sc-furwcr gPiovd"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-pVTFL yzYaz sc-jrQzAO gCBRHg"
      >
        <div
          class="sc-pVTFL ilAcon"
        >
          <label
            class="sc-ksdxgE fxGKky"
          >
            Foo
          </label>
          <svg
            class="sc-iqseJM eSaCCB"
            color="#FFFFFF"
            data-icon="cutArrowDown"
            height="2.4rem"
            isopen="0"
            style="margin-left: 1.7rem;"
            width="2.4rem"
          >
            cut-arrow-down.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="rah-static rah-static--height-zero"
      style="width: 100%; height: 0px; overflow: hidden;"
    >
      <div
        style="display: none;"
      >
        Bar
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`The Accordion should render correctly when open 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL sc-jRQBWg fGXmcv hJXOBd"
    style="width: 100%;"
  >
    <div
      aria-expanded="true"
      aria-label="open accordion"
      class="sc-furwcr gPiovd"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-pVTFL yzYaz sc-jrQzAO gCBRHg"
      >
        <div
          class="sc-pVTFL ilAcon"
        >
          <label
            class="sc-ksdxgE fxGKky"
          >
            Foo
          </label>
          <svg
            class="sc-iqseJM itDfL"
            color="#FFFFFF"
            data-icon="cutArrowDown"
            height="2.4rem"
            isopen="1"
            style="margin-left: 1.7rem;"
            width="2.4rem"
          >
            cut-arrow-down.svg
          </svg>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="rah-animating rah-animating--down rah-animating--to-height-auto"
      style="width: 100%; height: 0px; overflow: hidden; transition: height 500ms ease 0ms;"
    >
      <div
        style=""
      >
        Bar
      </div>
    </div>
  </div>
</DocumentFragment>
`;
